crates/compose/deny-all/src/fermyon.rs

```diff
@@ -0,0 +1,218 @@
+use crate::bindings::exports::fermyon::spin::{
+    key_value, 
+    llm, 
+    mqtt, 
+    mysql, 
+    postgres,
+    redis,
+    sqlite,
+    variables,
+};
+use crate::format_deny_error;
+
+pub struct KeyValueStore;
+
+impl key_value::GuestStore for KeyValueStore {
+    fn open(_label: String) -> Result<key_value::Store, key_value::Error> {
+        Err(key_value::Error::NoSuchStore)
+    }
+
+    fn get(&self, _key: String) -> Result<Option<Vec<u8>>, key_value::Error> {
+        unreachable!()
+    }
+
+    fn set(&self, _key: String, _value: Vec<u8>) -> Result<(), key_value::Error> {
+        unreachable!()
+    }
+
+    fn delete(&self, _key: String) -> Result<(), key_value::Error> {
+        unreachable!()
+    }
+
+    fn exists(&self, _key: String) -> Result<bool, key_value::Error> {
+        unreachable!()
+    }
+
+    fn get_keys(&self) -> Result<Vec<String>, key_value::Error> {
+        unreachable!()
+    }
+}
+
+impl key_value::Guest for crate::Component {
+    type Store = KeyValueStore;
+}
+
+impl llm::Guest for crate::Component {
+    fn infer(
+        _model: llm::InferencingModel,
+        _prompt: String,
+        _params: Option<llm::InferencingParams>,
+    ) -> Result<llm::InferencingResult, llm::Error> {
+        Err(llm::Error::ModelNotSupported)
+    }
+
+    fn generate_embeddings(
+        _model: llm::EmbeddingModel,
+        _text: Vec<String>,
+    ) -> Result<llm::EmbeddingsResult, llm::Error> {
+        Err(llm::Error::ModelNotSupported)
+    }
+}
+
+impl mqtt::GuestConnection for crate::Component {
+    fn open(
+        _address: String,
+        _username: String,
+        _password: String,
+        _keep_alive_interval_in_secs: u64,
+    ) -> Result<mqtt::Connection, mqtt::Error> {
+        Err(mqtt::Error::Other(format_deny_error("fermyon:spin/mqtt")))
```

Comment:
I'm not sure i agree. To me there is a useful distinction between the capability being disabled via virtualization and a misconfiguration of the otherwise enabled capability.

Reply:
Yes, but that distinction is useful to the operator, emitted via logs.  I'm not sure the component should be told.  But probably I am overly paranoid about what sort of things could lead to "Towlson in disclosure vulnerability shocker" on the HN front page...

Reply:
I think it _could_ be useful to the developer to provide fallback behavior and such. If i can't distinguish between "disabled" and "misconfigured" in code then there's no way for me to, e.g. gracefully fallback to using the local-randomly-generated-animal-fact-generator if outbound-http is disabled.

Reply:
I think we will have to agree to disagree.  It's not a blocker for sure!